glusterd2/transaction/transaction.go

```diff
@@ -13,91 +14,107 @@ import (
 	"github.com/coreos/etcd/clientv3/concurrency"
 	"github.com/pborman/uuid"
 	log "github.com/sirupsen/logrus"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/codes"
 )
 
 const (
-	txnPrefix = "transaction/"
+	txnPrefix  = "transaction/"
+	txnTimeOut = time.Second * 15
 )
 
-var expTxn = expvar.NewMap("txn")
+// TxnOptFunc receives a Txn and overrides its members
+type TxnOptFunc func(*Txn) error
 
 // Txn is a set of steps
 type Txn struct {
-	id          uuid.UUID
-	locks       Locks
-	reqID       uuid.UUID
-	storePrefix string
-
-	Ctx             TxnCtx
-	Steps           []*Step
-	DontCheckAlive  bool
-	DisableRollback bool
+	locks Locks
+
 	// Nodes is the union of the all the TxnStep.Nodes and is implicitly
 	// set in Txn.Do(). This list is used to determine liveness of the
 	// nodes before running the transaction steps.
-	Nodes   []uuid.UUID
-	OrigCtx context.Context
+	Nodes           []uuid.UUID `json:"nodes"`
+	StorePrefix     string      `json:"store_prefix"`
+	ID              uuid.UUID   `json:"id"`
+	Locks           []string    `json:"locks"`
+	ReqID           uuid.UUID   `json:"req_id"`
+	Ctx             TxnCtx      `json:"ctx"`
+	Steps           []*Step     `json:"steps"`
+	DontCheckAlive  bool        `json:"dont_check_alive"`
+	DisableRollback bool        `json:"disable_rollback"`
+	Initiator       uuid.UUID   `json:"initiator"`
+	StartTime       time.Time   `json:"start_time"`
+
+	success   chan struct{}
+	error     chan error
+	succeeded bool
 }
 
 // NewTxn returns an initialized Txn without any steps
 func NewTxn(ctx context.Context) *Txn {
 	t := new(Txn)
 
-	t.id = uuid.NewRandom()
-	t.reqID = gdctx.GetReqID(ctx)
+	t.ID = uuid.NewRandom()
+	t.ReqID = gdctx.GetReqID(ctx)
 	t.locks = make(map[string]*concurrency.Mutex)
-	t.storePrefix = txnPrefix + t.id.String() + "/"
+	t.StorePrefix = txnPrefix + t.ID.String() + "/"
 	config := &TxnCtxConfig{
 		LogFields: log.Fields{
-			"txnid": t.id.String(),
-			"reqid": t.reqID.String(),
+			"txnid": t.ID.String(),
+			"reqid": t.ReqID.String(),
 		},
-		StorePrefix: t.storePrefix,
+		StorePrefix: t.StorePrefix,
 	}
-	t.Ctx = newCtx(config)
-
-	t.OrigCtx = ctx
+	t.Ctx = NewCtx(config)
+	t.Initiator = gdctx.MyUUID
 	t.Ctx.Logger().Debug("new transaction created")
 	return t
 }
 
 // NewTxnWithLocks returns an empty Txn with locks obtained on given lockIDs
 func NewTxnWithLocks(ctx context.Context, lockIDs ...string) (*Txn, error) {
 	t := NewTxn(ctx)
+	t.Locks = lockIDs
+	return t, nil
```

Comment:
do we need to return an error?

Reply:
@Madhu-1 then we have to make changes in all files where this function has been called. Maybe we can take this up later.